README
Project EasyBashGUI.


Intro.

EasyBashGUI is a bash function library that aims to give simple GUI functions using yad, gtkdialog, kdialog, zenity, Xdialog, dialog, or whiptail depending on KDE or GNOME running or not, Yad, Gtkdialog or Xdialog installed or not and X server running or not ( (c)dialog or whiptail is the minimum ). So, if bash programmer writes: "message 'Thanks for using this program' ", he has not to worry in which environment his program runs: it is enough there is at least (c)dialog or whiptail installed, and program will work as expected. Obviously, if user has KDE, GNOME, or gtkdialog/Xdialog also installed ( and he's in a X session ), program will have "an other look", but logical flow is exactly the same.
You can force widget usage through "supermode" variable ( possible values: "gtkdialog", "kdialog", "zenity", "Xdialog", "dialog", "none" ; 
e.g.: >export supermode="kdialog" && source easybashgui && message Hello ).
NOTE on "none" mode: you have to make sure STDERR *is not* redirected, in order to see shell boxes in your terminal.

In EasyBashGUI "suite" there are a launcher ( "easybashgui" ), a launcher that toggles some debug options ( "easybashgui-debug" ), a widget library ( "easybashgui_X.X.X.lib" ), a script to quick test it ( "easybashgui_test.sh" ), and a stand-alone script to create dialog boxes externally ( "easydialog.sh" ). Moreover there is an other library too ( "easybashlib" ) for optional ancillary functions ( thanks to it you don't need anymore to use "clean_temp" function at the end of your scripts ).







Library functions.

message
ok_message
alert_message
text
question
input
menu
tagged_menu
list
fselect (= "file select" )
dselect (= "directory select" )
wait_seconds
wait_for
terminate_wait_for
progress
adjust






How to install and use.

Extract all files...
(e.g.: >tar -xzvf EasyBashGUI_X.X.X )

Cd in ./EasyBashGUI source dir...
(e.g.: >cd ./EasyBashGUI_X.X.X )

Install it (as root)...
(e.g.: >sudo make install )
 
Uninstall it (as root)...
(e.g.: >sudo make uninstall )

That's it !!!!!



If you want use it in your scripts, simply source "easybashgui" before use...
(e.g.: "source easybashgui" )

IMPORTANT: If easybashlib is present and successfully loaded, you can avoid to launch "clean_temp" to remove temporary files; 
           otherwise DO NOT forget to write "clean_temp" at the end of all your scripts... ;-)

E.g.:
-------------------
#!/bin/bash
source easybashgui
#
message "this"
input 1 ( "that" )
menu "this" "that"
list +"you" -"me" +"her"
...
etc.etc.
...
clean_temp #(since v.6.X.X this function is no more required if easybashlib is present and then successfully loaded by easybashgui)
-------------------

And... enjoy !! :-)

NOTE: Since v.5.X.X is possible making easybashgui work even if it's not installed in your system, 
      it's enough that your script and my libraries are all in the same directory.
      (e.g.: >cd /my_script/and/libraries/dir ; ./my_script )





Synopsis.

question -> "[text]"                 =>       ( 1 argument, box output to exit code and STDERR ) (*)
message -> "[text]"                  =>       ( 1 argument )
alert_message -> "[text]"            =>       ( 1 argument )
ok_message -> "[text]"               =>       ( 1 argument )
text                                 =>       ( STDIN, NO argument, box output to "${dir_tmp}/${file_tmp}" and STDERR ) (^)
wait_seconds -> "[integer]"          =>       ( 1 argument )
wait_for -> "[text]"                 =>       ( 1 argument, PID to kill to "wait_for__PID" variable and STDERR ) (@)
terminate_wait_for                   =>       ( 1 argument only in easydialog, otherwise, NO argument ) (@)
fselect -> "<init. dir.>"            =>       ( 1 <optional> argument, box output to "${dir_tmp}/${file_tmp}" and STDERR ) (#)
dselect -> "<init. dir.>"            =>       ( 1 <optional> argument, box output to "${dir_tmp}/${file_tmp}" and STDERR ) (#)
input -> 1 "[init 1]"                =>       ( 2 arguments, box output to "${dir_tmp}/${file_tmp}" and STDERR )
input -> 2 "[label 1]" "[init 1]" "[label 2]" "[init 2]"                     => ( 5 arguments, box output to "${dir_tmp}/${file_tmp}" and STDERR )
input -> 3 "[label 1]" "[init 1]" "[label 2]" "[init 2]" "[label 3]" "[init 3]" => ( 7 arguments, box output to "${dir_tmp}/${file_tmp}" and STDERR )
menu -> "[item 1]" ... "[item n]"    =>       ( [n] arguments, box output to "${dir_tmp}/${file_tmp}" and STDERR ) (%)
tagged_menu -> "[tag 1]" "[item 1]" ... "[tag n]" "[item n]"    =>       ( [n*2] arguments, box output to "${dir_tmp}/${file_tmp}" and STDERR ) (%)
list -> <+|->"[item 1]" ... <+|->"[item n]"    =>       ( [n] arguments, optionally prefixed by "+"(plus) or "-"(minus), box output to "${dir_tmp}/${file_tmp}" and STDERR ) (%)
progress -> "[text]"                         =>       ( percent with or without '%' in STDIN, 1 argument )
progress -> "[text]" "[elements number]"     =>       ( "PROGRESS" string in STDIN, 2 arguments )
adjust -> "[text]" "[min]" "[init]" "[max]" => ( 4 arguments, box output to "${dir_tmp}/${file_tmp}" and STDERR )


(*) = "0" exit status is "YES", "1" exit status is "NOT", other exit codes you should make program exit : normally in a script you have just to check exit status to know user choice ;
(^) = text function write text in STDIN to file "${dir_tmp}/${file_tmp}" and (only) for kdialog, zenity, and Xdialog you can also edit text to write ;
(@) = "wait_for" function create a window with a text and returns control to main program... after a job, you can close the window throught function "terminate_wait_for" ( needs no argument ) ;
(#) = take care that if you are in "console mode" or without X, throught cdialog, selection is done by SPACE key, and NOT by enter key : remember it ;
(%) = "menu" and "list" functions differ about choices: menu allows single choice, list allows multiple choice ; since 7.1.0 version you can use tagged_menu(): it outputs tags (e.g.: "tagged_menu 1 A 2 B" -> if user selects "A" then function outputs "1" );

Since EasyBashGUI v.1.2.4, all windows functions support options "<-w|-width> [integer]", and "<-h|-height> [integer]" for custom window size:
E.g.: >alert_message -w 400 -h 340 "Error!"



Library examples:
1)
question "Do you like Contry music ?"
answer="${?}"
if [ ${answer} -eq 0 ]
	then
	ok_message "You do like it :)"
elif [ ${answer} -eq 1 ]
	then
	alert_message "You don't like it :("
else
	ok_message "See you"
	exit 0
fi

2)
echo -e "What's your name?\n\nMy name's:\nVittorio" | text

3)
wait_for "I'm sleeping 4 seconds... good night..."
sleep 4
terminate_wait_for

4)
fselect
file="$(0< "${dir_tmp}/${file_tmp}" )"

5)
input 1 "(write here IP address)"
input 3 "Username" "root" "IP address" "192.168.0.1" "Destination directory" "/tmp"
IFS=$'\n' ; choices=( $(0< "${dir_tmp}/${file_tmp}" ) ) ; IFS=$' \t\n'
user="${choices[0]}"
ip="${choices[1]}"
dir="${choices[2]}"

6)
for i in 10 20 30 40 50 60 70 80 90 100
	do
	echo "${i}"
	sleep 2
done | progress "This is a test progress..."

7)
adjust "Please, set Volume level" 15 40 75

8)
women=( Angela Carla Michelle Noemi Urma Marisa Karina Anita Josephine Rachel )
for (( index=0 ; index < ${#women[@]} ; index++ })) 
	do
	today_prefered_woman="${women[${index}]}"
	kiss "${today_prefered_woman}"
	sleep 1
	#
	# Job done !!
	# then...
	echo "PROGRESS"
	#
done | progress "This is a _LOVE_ progress..." "${#women[@]}"
# if you use "PROGRESS" string in STDIN do not forget second argument ( "[elements number]" )

( For easydialog.sh use and examples, you would launch it simply with "-h" option )







Note on console mode.

EasyBashGUI doesn't work with original "dialog" ( old one ) that is very limited; if you have first version "dialog" in your box, install "cdialog" and alias or link "dialog" to cdialog. No problem in case you have at least "whiptail" installed: since version 4.0.0, EasyBashGUI is able to use it instead of (c)dialog.
Since 5.0.0 version you can use EasyBashGUI even if NO WIDGET is installed (that is: no gtkdialog, no kdialog, no zenity, no Xdialog, no (c)dialog, no whiptail... doh!!!!! ). To use "super bare" EBG, simply remove the ".lib" library from your path, or set "supermode" var to "none" before easybashgui sourcing (e.g.: >export supermode="none" && source easybashgui && message "Hello world..." )





Note on gtkdialog mode.

EasyBashGUI sets gtkdialog output statements as variables through "eval". This way, in theory, could be possibly dangerous; nevertheless, so far, I don't know about an alternative way...





Note on Git.
Since 8.0.0 version, project is housed at github ( https://github.com/BashGui )



Credits.

Thanks to Bash, Yad, Gtkdialog, Xdialog, Kdialog, Zenity, Cdialog, and Whiptail authors, this library was nothing without their work. Many thanks.
Thanks to Jose Joao Dias de Almeida for the makefile tip.
Thanks to Chris "cgat" for his many ideas and suggestions that lead to version's "5.X.X" EasyBashGUI "revolution".
Thanks to Davide Depau for his tests and support, and his effort to make EasyBashGUI Debian policy compliant, and finally for his man page.
Thanks to Christian Prause for his patience and Git support.
Thanks to GitHub for housing our project.
Thanks to Lucio Messina for "Debian policy" tips and support. 







Please let me know if my work was useful for you.
Vittorio Cagnetta
vaisarger@gmail.com
https://github.com/BashGui/easybashgui


debian/changelog
easybashgui (8.0.1~precise-1) precise; urgency=low

  * Initial release: (Closes: "Create RPMs, DEBs, etc")

 -- Lucio Messina <lucio.messina@mensa.it>  Sat, 06 Jul 2013 09:20:18 +0200


easybashgui_test.sh
#!/usr/bin/env bash
#
#########################
#
# Copyright (C) 2011 Vittorio Cagnetta
#
# Author: Vittorio Cagnetta <vaisarger@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with this program, 
# called, in this distribution, "EasyBashGUI-license"; if not, write to the 
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
#
#########################
#
source easybashgui
#
LIB_CHECK="$(type "easybashgui" 2> /dev/null )"
[ ${#LIB_CHECK} -eq 0 ] && echo -e "\n\n\n\nYou need to copy \"easybashgui\" in your path ( e.g.: \"cp easybashgui_X.X.X /usr/local/bin/\" )...\n:(\n\n\n\n" 1>&2 && exit 1
#
##
#
##################
show()
	{
	message "You chose/wrote: \n\n \"$(cat "${dir_tmp}/${file_tmp}" )\" \n\n ;-)" 
	}
##################
#
##
#
# ... "x" is the function name...
#
##
#
x=message,ok_message,alert_message
	{
	x=message
	${x} "This is a test \"<${x}>\" function...\n;-)"
	#
	x="ok_message --width 700 --height 300"
	[ "${mode}" = "kdialog" ] && message "Now let's try to change \n window size... \n\n Next function: \"<${x}>\" ... \n\n ( Note: for kdialog messages, \n - due to a known kdialog bug - \n setting a different geometry \n makes it ONLY fit to box text !! )" || \
	message "Now let's try to change \n window size... \n\n Next function: \"<${x}>\" ..."
	${x} "This is a test \"<${x}>\" function, \n with custom size... \n ;-)"
	#
	x="alert_message --height 700 --width 300"
	[ "${mode}" = "kdialog" ] && message "Now let's try to change \n window size... \n\n Next function: \"<${x}>\" ... \n\n ( Note: for kdialog messages, \n - due to a known kdialog bug - \n setting a different geometry \n makes it ONLY fit to box text !! )" || \
	message "Now let's try to change \n window size... \n\n Next function: \"<${x}>\" ..."
	${x} "This is a test \"<${x}>\" function, \n with custom size... \n ;-)"
	}
#
##
#
x=text
	{
	file="/etc/passwd"
	message "Next function: \"head -n 10 ${file} | ${x} -w 600 -h 600\" ..."
	head -n 10 ${file} | ${x} -w 600 -h 600 ; show
	}
#
##
#
x=question
	{
	while :
		do
		${x} "This is a test \"<${x}>\" function (1)... \n Please click \"Cancel\", now ..."
		answer="${?}"
		if [ ${answer} -eq 0 ]
			then
			continue
		elif [ ${answer} -eq 1 ]
			then
			break
		else
			exit
		fi
	done
	while :
		do
		${x} "This is a test \"<${x}>\" function (2)... \n Please click \"Ok\", this time ..."
		answer="${?}"
		if [ ${answer} -eq 0 ]
			then
			break
		elif [ ${answer} -eq 1 ]
			then
			continue
		else
			exit
		fi
	done
	}
#
##
#
x=wait_seconds
	{
	message "Next function: \"<${x} 5>\" ..."
	${x} 5
	}
#
##
#
x=wait_for
	{
	message "Next function: \"<${x}>\" ..."
	${x} "This is a test ${x} function (command: sleep 5) \n + \n terminate_wait_for \n\n ;-)"
	sleep 5
	terminate_wait_for
	}
#
##
#
x=fselect,dselect
	{
	for x in fselect dselect
		do
		message "Next function: \"<${x}>\" ..."
		${x} ; show
	done
	}
#
##
#
x=input
	{
	message "Next function: \"<${x} 1>\" ..."
	${x} 1 "(please write here)" ; show
	#
	message "Next function: \"<${x} 2>\" ..."
	${x} 2 "First name" "Jimi" "Surname" "Hendrix" ; show
	#
	message "Next function: \"<${x} 3>\" ..."
	${x} 3 "Artist" "Albert King" "Genre" "Blues" "Album" "The Big Blues" ; show
	}
#
##
#
x=menu
	{
	x="menu -w 650 -h 550"
	message "Next function: \"<${x}>\" ..."
	${x} "Country Music" "Blues" "Rock And Roll" "Pop" ; show
	}
#
##
#
x=list
	{
	message "Next function: \"<${x}>\" ... \n ( it differs from \"menu\" for multi-choice possibility )"
	list -"Country Music" +"Blues" +"Rock And Roll" -"Pop" ; show
	}
#
##
#
x=progress
	{
	x="progress -w 600"
	message "Next function: \"${x}\" with \"custom\" width ( \"-w\" ) \n (old style with \n percent integer -with or without '%'- \n in STDIN)..."
	for i in 10 20.30 32 40.50 50 60.50 70.45 80.90 85 93 100; do echo "${i}"; sleep 1 ; done | ${x} "This is a test \"<${x}>\" function \n (old style with percent in STDIN)... \n\n ;-)"
	#
	x="progress --width 400 --height 200"
	message "Next function: \"${x}\" with \"custom\" width and height \n ( \"--width\" \"--height\" ) \n (new style with \n \"PROGRESS\" string \n in STDIN)..."
	for i in 10 20.30 32 40.50 50 60.50 70.45 80.90 85 93 100; do echo "PROGRESS" ; sleep 1 ; done | ${x} "This is a test \"<${x}>\" function \n (new style with \"PROGRESS\" string in STDIN)... \n\n ;-)" 11
	}
#
##
#
x=adjust
	{
	message "Next function: \"<${x} 'Please adjust level...' 100 200 300>\" ..."
	${x} 'Please adjust level...' 100 200 300 ; show
	}
#
##
#
ok_message "Tour is finished. :-) \n Click now to see with your eyes \n how easy is programming with ${LIB_NAME}..."
#
##
#
# This is *short* and *nice* :
cat "${0}" | text -w 800 -h 600
#
sleep 3 && clean_temp





easydialog.sh
#!/usr/bin/env bash
#
#########################
#
# Copyright (C) 2013 Vittorio Cagnetta
#
# Author: Vittorio Cagnetta <vaisarger@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with this program, 
# called, in this distribution, "EasyBashGUI-license"; if not, write to the 
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
#
#########################
#
############################
#
#
#
usage()
	{
	#
	echo -e "Usage:>$(basename "${0}" ) [-m [mode]] [box type] <box options>" 
	echo -e "\t1)Where [modes] are:" 
	echo -e "\t a \t\t=>\t automatic (normal way)"
	echo -e "\t g \t\t=>\t to force gtkdialog"
	echo -e "\t k \t\t=>\t to force kdialog"
	echo -e "\t z \t\t=>\t to force zenity"
	echo -e "\t x \t\t=>\t to force Xdialog"
	echo -e "\t d \t\t=>\t to force (c)dialog"
	echo -e "\t n \t\t=>\t to force no widget ( => shell boxes)"
	echo -e "\n"
	echo -e "\tAnd [box types] -> [box options] are:" 
	echo -e "\t --question -> \"[text]\" \t\t=>\t ( 1 argument, box output to exit code and STDERR )"
	echo -e "\t --message -> \"[text]\" \t\t\t=>\t ( 1 argument )"
	echo -e "\t --alert_message -> \"[text]\" \t\t=>\t ( 1 argument )"
	echo -e "\t --ok_message -> \"[text]\" \t\t=>\t ( 1 argument )"
	echo -e "\t --text \t\t\t\t=>\t ( STDIN, NO argument, box output to \"\${dir_tmp}/\${file_tmp}\" and STDERR )"
	echo -e "\t --wait_seconds -> \"[integer]\" \t\t=>\t ( 1 argument )"
	echo -e "\t --wait_for -> \"[text]\" \t\t=>\t ( 1 argument, PID to kill to \"wait_for__PID\" variable and STDERR )"
	echo -e "\t --terminate_wait_for \t\t\t=>\t ( 1 argument only in $(basename "${0}" ), otherwise, NO argument )"
	echo -e "\t --fselect -> \"<init. dir.>\" \t\t=>\t ( 1 <optional> argument, box output to \"\${dir_tmp}/\${file_tmp}\" and STDERR )"
	echo -e "\t --dselect -> \"<init. dir.>\" \t\t=>\t ( 1 <optional> argument, box output to \"\${dir_tmp}/\${file_tmp}\" and STDERR )"
	echo -e "\t --input -> 1 \"[init 1]\" \t\t=>\t ( 2 arguments, box output to \"\${dir_tmp}/\${file_tmp}\" and STDERR )"
	echo -e "\t --input -> 2 \"[label 1]\" \"[init 1]\" \"[label 2]\" \"[init 2]\" \t\t\t=> ( 5 arguments, box output to \"\${dir_tmp}/\${file_tmp}\" and STDERR )"
	echo -e "\t --input -> 3 \"[label 1]\" \"[init 1]\" \"[label 2]\" \"[init 2]\" \"[label 3]\" \"[init 3]\" => ( 7 arguments, box output to \"\${dir_tmp}/\${file_tmp}\" and STDERR )"
	echo -e "\t --menu -> \"[item 1]\" ... \"[item n]\" \t=>\t ( [n] arguments, box output to \"\${dir_tmp}/\${file_tmp}\" and STDERR )"
	echo -e "\t --list -> <+|->\"[item 1]\" ... <+|->\"[item n]\" \t=>\t ( [n] arguments, optionally prefixed with \"+\"(plus) or \"-\"(minus), box output to \"\${dir_tmp}/\${file_tmp}\" and STDERR )"
	echo -e "\t --progress -> \"[text]\" \t\t\t=>\t ( STDIN -percent-, 1 argument )"
	echo -e "\t --progress -> \"[text]\" \"[elements number]\" \t=>\t ( STDIN -\"PROGRESS\"-, 2 arguments )"
	echo -e "\t --adjust -> \"[text]\" \"[min value]\" \"[init value]\" \"[max value]\" \t=>\t ( 4 arguments, box output to \"\${dir_tmp}/\${file_tmp}\" and STDERR )"
	echo -e "\n\n"
	echo -e "\t2)Examples:"
	echo -e "\t>$(basename "${0}" ) -m a --question \"Do you like Contry music ?\" 2> a.txt ; a=\"\$(0< a.txt )\" ; rm -f a.txt ; [ \${a} -eq 0 ] && $(basename "${0}" ) -m a --ok_message \"You do like it :)\" || $(basename "${0}" ) -m a --alert_message \"You don't like it :(\" "
	echo -e "\t>echo -e \"What's your name?\\\n\\\nMy name's:\\\nVittorio\" | $(basename "${0}" ) -m z --text"
	echo -e "\t>$(basename "${0}" ) -m a --wait_seconds 3"
	echo -e "\t>$(basename "${0}" ) -m k --wait_for \"I'm sleeping 4 seconds... good night...\" 2> pid.txt ; pid=\"\$(0< pid.txt )\" ; rm -f pid.txt ;    sleep 4    ; $(basename "${0}" ) -m k --terminate_wait_for \"\${pid}\" "
	echo -e "\t>$(basename "${0}" ) -m x --fselect"
	echo -e "\t>$(basename "${0}" ) -m a --input 1 \"(write here IP address)\""
	echo -e "\t>$(basename "${0}" ) -m k --input 3 \"Username\" \"root\" \"IP address\" \"192.168.0.1\" \"Destination directory\" \"/tmp\""
	echo -e "\t>$(basename "${0}" ) -m d --menu \"Country\" \"Blues\" \"Rock\""
	echo -e "\t>for i in 10 20 30 40 50 60 70 80 90 100 ; do echo \"\${i}\" && sleep 2 ; done | $(basename "${0}" ) -m d --progress \"This is a test progress...\""
	echo -e "\t>for i in 10 20 30 40 50 60 70 80 90 100 ; do echo \"PROGRESS\" && sleep 2 ; done | $(basename "${0}" ) -m z --progress \"This is a test progress...\" \"10\""
	echo -e "\n\n"
	echo -e "\t3)Would you like *really* easy dialogs?"
	echo -e "\tTry this:"
	echo -e "\t>cp easybashgui /usr/local/bin/"
	echo -e "\t>touch \"easy_updatedb.sh\""
	echo -e "\t>chmod 755 \"easy_updatedb.sh\""
	echo -e "\tThen, edit the file like so:"
	echo -e "\t"
	echo -e "\t#!/usr/bin/env bash"
	echo -e "\tsource /usr/local/bin/easybashgui"
	echo -e "\tmessage \"Click OK \\\n to update locate db...\""
	echo -e "\twait_for \"Updating...\" ; updatedb ; terminate_wait_for"
	echo -e "\tok_message \"Update finished ;-)\""
	echo -e "\t"
	echo -e "\tTry to launch the *same* script in console, in GNOME, in KDE, and finally in text console... ;-)"
	echo
	#
	exit 1
	#
	} 1>&2
#
#
#
############################
#
#
#
[ ${#} -eq 0 ] && usage
while getopts ":hvm:" option
	do
	case ${option} in
	 "h" )
		source easybashgui ; \
		echo -e "\nThe library \"${LIB_NAME}\" allows bash scripters to simply provide their scripts a graphical user interface in an easy way: no matter wich dialog type is more appropriate ( if you are in GNOME, in KDE, etc.).\nThe script ( \"$(basename "${0}" )\" ) is used only to show ${LIB_NAME} library capability...\n\n" 1>&2 && usage \
		;;
	 "v" )
		source easybashgui ; \
		echo -e "\n${LIB_NAME}: vers. ${LIB_VERSION}\n" 1>&2 && exit \
		;;
	 "m" )
		if [ "${OPTARG}" = "a" ]; then export supermode="auto" && break ; \
		elif [ "${OPTARG}" = "g" ]; then export supermode="gtkdialog" && break ; \
		elif [ "${OPTARG}" = "k" ]; then export supermode="kdialog" && break ; \
		elif [ "${OPTARG}" = "z" ]; then export supermode="zenity" && break ; \
		elif [ "${OPTARG}" = "x" ]; then export supermode="Xdialog" && break ; \
		elif [ "${OPTARG}" = "d" ]; then export supermode="dialog" && break ; \
		elif [ "${OPTARG}" = "n" ]; then export supermode="none" && break ; \
		else usage ; \
		fi \
		;;
	 * )
		usage \
		;;
	esac
	#
done
shift $(($OPTIND - 1))
#
#
#
funct="$(echo "${1}" | tr -d '-' )"
case ${funct} in
	"question"|\
	"message"|"alert_message"|"ok_message"|\
	"text"|\
	"wait_seconds"|"wait_for"|"terminate_wait_for"|\
	"fselect"|"dselect"|\
	"input"|\
	"menu"|"list"|\
	"progress"|\
	"adjust" )
		:;;
	
	* )
		usage ;;
esac && shift
#
#
#
############################
# ...main
LIB_CHECK="$(type "easybashgui" 2> /dev/null )" ; [ ${#LIB_CHECK} -eq 0 ] && echo -e "\n\n\n\nYou need to copy \"easybashgui\" in your path ( e.g.: \"cp easybashgui_X.X.X /usr/local/bin/\" )...\n:(\n\n\n\n" 1>&2 && usage
#
source easybashgui
num_param=${#}
eval ${funct} $(for (( index=1 ; index <= ${num_param} ; index++ ))
			do
			echo -n "\"${1}\" "
			shift
		done ) 
clean_temp
# bye! ;-)
